app/assets/javascripts/orders.coffee

```diff
@@ -41,74 +41,78 @@ app.factory('Address', ['$resource', ($resource) ->
     }
 ])
 
-app.controller 'ProductsController', ['$scope', 'Product', 'OrderItem', 'Order', '$window', ($scope, Product, OrderItem, Order, $window) ->
-  cartPromise = Order.cart().$promise
+app.controller 'BaseController', ['$scope', 'Order', 'OrderItem', '$window', ($scope, Order, OrderItem, $window) ->
+  Order.cart().$promise.then (data) ->
+      if data?
+        $scope.order = new Order(data.order)
+      else
+        $scope.order = new Order()
 
-  cartPromise.then (data) ->
-    if data?
-      $scope.order = new Order(data.order)
-    else
-      $scope.order = new Order()
+  $scope.addToCart = (item) ->
+      return if $scope.addingToCart
+      $scope.addingToCart = true
+
+      existingItem = _.findWhere($scope.order.order_items, { source_id: item.id })
+      if existingItem?
+        existingItem.quantity = item.quantity
+        itemToSave = new OrderItem(existingItem)
+        promise = itemToSave.$update()
+      else
+        itemToSave = new OrderItem(source_id: item.id, quantity: item.quantity, source_type: "Product")
+        itemToSave.order_id = $scope.order.id
+        promise = itemToSave.$save()
+
+      promise
+        .then (orderItem) ->
+          $scope.order.order_items.push(orderItem.order_item) unless existingItem
+          $scope.saveSuccess = true
+        .catch (result) ->
+          $scope.productErrors = result.order_items
+        .finally ->
+          $scope.addingToCart = false
+
+   $scope.removeFromCart = (item, index) ->
+    $scope.order.order_items.splice(index, 1)
 
-    $scope.order.order_items.map (item) ->
-      new OrderItem(item)
+    itemToDelete = new OrderItem(item)
+    itemToDelete.$delete()
+]
+
+app.controller 'ProductsController', ['$scope', 'Product', 'OrderItem', 'Order', '$window', ($scope, Product, OrderItem, Order, $window) ->
 
   Product.get().$promise.then (data) ->
       $scope.products = data.products
-
-  $scope.addToCart = (item) ->
-    cartPromise.then (data) ->
-      return if _.findWhere($scope.order.order_items, { source_id: item.id })
-      ItemToSave = new OrderItem(source_id: item.id, quantity: item.quantity, source_type: "Product")
-      ItemToSave.order_id = $scope.order.id
-
-      ItemToSave.$save()
-      .then ->
-        $scope.saveSuccess = true
-      .catch (result) ->
-        $scope.productErrors = result.errors.product
 ]
 
-app.controller 'CheckoutController',['$scope', 'Product', 'Order', 'Address', 'CreditCard', 'OrderItem', ($scope, Product, Order, Address, CreditCard, OrderItem) ->
+app.controller 'CheckoutController',['$q', '$scope', 'Product', 'Order', 'Address', 'CreditCard', 'OrderItem', ($q, $scope, Product, Order, Address, CreditCard, OrderItem) ->
 
   Order.cart().$promise.then (data) ->
     if data?
-      $scope.order = new Order(data.order)
+      $scope.ord = new Order(data.order)
```

Comment:
I know you're using ```$scope.ord``` instead of ```$scope.order``` because there is another controller that sets ```$scope.order``` but are you actually using the ```$scope.ord``` in the view anywhere?